### PR TITLE
571 Generalize ALP Matrix and Vector over backend

### PR DESCRIPTION
**Pull request by vdimiczrc**

### CODE REVIEW

File: include/alp/reference/matrix.hpp [hunk 1]
```diff
@@ -579,16 +491,16 @@ namespace alp {
...
 
-			protected:
-				typedef StorageBasedMatrix< T, AmfType, requires_allocation > self_type;
-				friend MatrixBase< self_type >;
-
 				typedef typename std::conditional<
 					requires_allocation,
 					Vector< T, reference >,
 					Vector< T, reference > &
 				>::type container_type;
```
Comment (hidanielesgit) [inline include/alp/reference/matrix.hpp:498]: Why is this becoming public?

File: include/alp/reference/matrix.hpp [hunk 2]
```diff
@@ -878,516 +773,16 @@ namespace alp {
...
-				static_assert(
-					std::is_same< ImfR, imf::Id >::value &&
-					std::is_same< ImfC, imf::Id >::value,
-					"This constructor can only be used with Id IMFs."
-				);
-
-			}
-	}; // ALP Matrix
-
 	namespace structures {
```
Comment (hidanielesgit) [inline include/alp/reference/matrix.hpp:784]: Perhaps should go to structure.hpp?

File: include/alp/reference/matrix.hpp [hunk 3]
```diff
@@ -878,516 +773,16 @@ namespace alp {
...
-			static_assert( std::is_base_of< structures::BaseStructure, TestedStructure >::value );
-
-			/**
-			 * \a value is true iff \a Structure is implied by \a TestedStructure.
-			 */
-			static constexpr bool value = is_in< Structure, typename TestedStructure::inferred_structures >::value;
-		};
-
 		template<size_t band, typename T, typename Structure, enum Density density, typename View, typename ImfL, typename ImfR >
 		std::ptrdiff_t get_lower_limit(const alp::Matrix< T, Structure, density, View, ImfL, ImfR, reference > &A) {
```
Comment (hidanielesgit) [inline include/alp/reference/matrix.hpp:787]: Perhaps move this and others below to amf_based/matrix.hpp ?

File: include/alp/reference/matrix.hpp [hunk 4]
```diff
@@ -1518,412 +898,6 @@ namespace alp {
...
-		const SelectVectorC &sel_c
-	) {
-		return internal::get_view< TargetStructure >(
-			source,
-			imf::Select( nrows( source ), sel_r ),
-			imf::Select( ncols( source ), sel_c )
-		);
-	}
-
 	namespace structures {
```
Comment (hidanielesgit) [inline include/alp/reference/matrix.hpp:901]: Also this block to amf_based/matrix.hpp?

**Review by hidanielesgit — APPROVED** [review]

File: include/alp/amf-based/vector.hpp [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+
+			/** @see Vector::value_type. */
+			using value_type = T;
+
+			typedef structures::General structure;
+
+			/** @see Vector::lambda_reference */
+			typedef typename std::conditional<
+				internal::is_storage_based< self_type >::value,
+				T&,
```
Comment (djelovina) [inline include/alp/amf-based/vector.hpp:136]: space between `T` and `&`, here and on other places

**Review by djelovina — APPROVED** [review]

**Pull request merged**